Modules/posixmodule.c

```diff
@@ -13161,6 +13161,99 @@ os_getrandom_impl(PyObject *module, Py_ssize_t size, int flags)
 }
 #endif   /* HAVE_GETRANDOM_SYSCALL */
 
+#ifdef MS_WINDOWS
+/* bpo-36085: Helper functions for managing DLL search directories
+ * on win32
+ */
+
+typedef DLL_DIRECTORY_COOKIE (*PAddDllDirectory)(PCWSTR newDirectory);
+typedef BOOL (*PRemoveDllDirectory)(DLL_DIRECTORY_COOKIE cookie);
+
+/*[clinic input]
+os.add_dll_directory
+
+    path: path_t
+
+Add a path to the DLL search path.
+
+This search path is used when resolving dependencies for imported
+extension modules (the module itself is resolved through sys.path),
+and also by ctypes.
+
+Returns an opaque value that may be passed to os.remove_dll_directory
+to remove this directory from the search path.
+[clinic start generated code]*/
+
+static PyObject *
+os_add_dll_directory_impl(PyObject *module, path_t *path)
+/*[clinic end generated code: output=8e498e193ddce271 input=bae8b8ffda652600]*/
+{
+    HMODULE hKernel32;
+    PAddDllDirectory AddDllDirectory;
+    DLL_DIRECTORY_COOKIE cookie = 0;
+
+    /* For Windows 7, we have to load this. As this will be a fairly
+       infrequent operation, just do it each time. Kernel32 is always
+       loaded. */
+    Py_BEGIN_ALLOW_THREADS
+    if ((hKernel32 = GetModuleHandleW(L"kernel32")) &&
+        (AddDllDirectory = (PAddDllDirectory)GetProcAddress(
+            hKernel32, "AddDllDirectory"))) {
+        cookie = (*AddDllDirectory)(path->wide);
+    }
+    Py_END_ALLOW_THREADS
+
+    if (!cookie) {
+        return win32_error_object("AddDllDirectory", path->object);
+    }
+
+    return PyCapsule_New(cookie, "DLL directory cookie", NULL);
+}
+
+/*[clinic input]
+os.remove_dll_directory
+
+    cookie: object
+
+Removes a path from the DLL search path.
+
+The parameter is an opaque value that was returned from
+os.add_dll_directory. You can only remove directories that you added
+yourself.
+[clinic start generated code]*/
+
+static PyObject *
+os_remove_dll_directory_impl(PyObject *module, PyObject *cookie)
+/*[clinic end generated code: output=cfbb576332208c2a input=178939bb10c7a6cf]*/
+{
+    HMODULE hKernel32;
+    PRemoveDllDirectory RemoveDllDirectory;
+    DLL_DIRECTORY_COOKIE cookieValue;
+
+    if (!PyCapsule_IsValid(cookie, "DLL directory cookie")) {
```

Comment:
For anyone wondering, it's necessary to verify it's a valid cookie because it's an opaque pointer. Calling `RemoveDllDirectory` on an invalid cookie will attempt to read from the address and possibly terminate the process on an unhandled access violation.

Reply:
It certainly does terminate the process... :(